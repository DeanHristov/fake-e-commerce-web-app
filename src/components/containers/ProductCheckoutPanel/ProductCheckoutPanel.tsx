import { FC } from 'react';
import Widget from '@/components/containers/Widget';
import { Utils } from '@/utils/Utils';
import InputField from '@/components/ui/InputField';
import Button from '@/components/ui/Button';
import { TCurrency } from '@/types';
import { useRouter } from 'next/navigation';

export interface IProductCheckoutPanelProps {
  totalQuantity: number;
  totalProducts: number;
  total: number;
  discountedTotal: number;
  userId: number | null;
  currency: TCurrency;
}

const ProductCheckoutPanel: FC<IProductCheckoutPanelProps> = ({
  totalQuantity,
  totalProducts,
  total,
  userId,
  discountedTotal,
  currency,
}) => {
  const router = useRouter();
  return (
    <Widget className="max-h-[16rem] w-full lg:w-auto">
      <div className="p-2">
        {Utils.isNotNull(totalQuantity) && (
          <h3 className="px-3 py-3 text-center text-base text-black border-b border-b-dark">{`${total} Items in your Bag`}</h3>
        )}
        <p className="py-2 mb-4 border-b border-b-dark uppercase flex justify-between items-center">
          <span className={'font-bold text-md text-black'}>{`total `}</span>

          <span className={'font-medium text-md text-black'}>
            {Utils.isNull(discountedTotal)
              ? '---'
              : Utils.parseAmountByCurrency(discountedTotal, currency)}
          </span>
        </p>
        <InputField
          disabled={Utils.isNull(totalProducts)}
          placeholder="Enter a promocode"
          className="w-full h-10 mb-4 mx-auto max-w-xl"
          onChange={(ev) => console.log(ev.target.value)}
        />

        <Button
          disabled={Utils.isNull(total)}
          className="rounded-none"
          variant="primary"
          title={'Checkout'}
          onClick={() => router.push('/checkout')}
        />
      </div>
    </Widget>
  );
};

export default ProductCheckoutPanel;
