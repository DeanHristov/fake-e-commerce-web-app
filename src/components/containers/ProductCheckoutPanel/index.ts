import ProductCheckoutPanel, {
  IProductCheckoutPanelProps,
} from './ProductCheckoutPanel';

export type { IProductCheckoutPanelProps };
export default ProductCheckoutPanel;
