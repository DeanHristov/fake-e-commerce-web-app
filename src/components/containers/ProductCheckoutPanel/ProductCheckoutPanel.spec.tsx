import { render } from '@testing-library/react';
import ProductCheckoutPanel from './';
import userEvent from '@testing-library/user-event';

// TODO Basically, is a good idea rest of the use cases to be covered from automation tests!
describe('Container/ProductCheckoutPanel <ProductCheckoutPanel {...} />', () => {
  it('Should render a default state of the Component', () => {
    const { container, getByRole } = render(
      <ProductCheckoutPanel
        onCheckout={jest.fn}
        totalQuantity={0}
        totalProducts={0}
        total={0}
        discountedTotal={0}
        userId={1}
        currency={'EUR'}
      />,
    );

    expect(getByRole(`button`, { name: /Checkout/i })).toBeDisabled();
    expect(container).toMatchSnapshot();
  });

  it('Should be able to do checkout of selected orders', async () => {
    const mockCheckout = jest.fn();

    const { getByRole, getByText } = render(
      <ProductCheckoutPanel
        onCheckout={mockCheckout}
        totalQuantity={1}
        totalProducts={1}
        total={100}
        discountedTotal={100}
        userId={1}
        currency={'EUR'}
      />,
    );

    const $button: HTMLElement = getByRole('button', { name: /Checkout/i });
    await userEvent.click($button);

    expect(getByText('â‚¬100.00')).toBeVisible();
    expect(getByRole(`button`, { name: /Checkout/i })).not.toBeDisabled();
    expect(mockCheckout).toHaveBeenCalledWith(100);
  });
});
