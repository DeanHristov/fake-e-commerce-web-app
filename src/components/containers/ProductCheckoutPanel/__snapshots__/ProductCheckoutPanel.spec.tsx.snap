// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container/ProductCheckoutPanel <ProductCheckoutPanel {...} /> Should render a default state of the Component 1`] = `
<div>
  <div
    class="bg-white border border-gray-300 rounded-lg p-4 shadow-md inline-block relative max-h-[16rem] w-full lg:w-auto"
  >
    <div
      class="p-2"
    >
      <h3
        class="px-3 py-3 text-center text-base text-black border-b border-b-dark"
      >
        0 Items in your Bag
      </h3>
      <p
        class="py-2 mb-4 border-b border-b-dark uppercase flex justify-between items-center"
      >
        <span
          class="font-bold text-md text-black"
        >
          total 
        </span>
        <span
          class="font-medium text-md text-black"
        >
          ---
        </span>
      </p>
      <div
        class="relative w-full rounded-md w-full h-10 mb-4 mx-auto max-w-xl border border-gray-500"
      >
        <input
          class="h-full px-6 w-full rounded-lg bg-white text-slate-700 ease-in duration-200 focus:bg-white"
          disabled=""
          placeholder="Enter a promocode"
          type="text"
        />
      </div>
      <button
        class="cursor-not-allowed opacity-50 bg-green-600 text-white font-bold uppercase bg-primary hover:bg-hover hover:shadow-lg transition transform active:scale-90 duration-200 ease-in w-full px-4 py-1.5 flex justify-center items-center gap-2 rounded-none"
        disabled=""
      >
        Checkout
      </button>
    </div>
  </div>
</div>
`;
